add FaktoryWorker.{Sandbox, Testing} test utilities

superhawk610 opened the pull request:
This PR introduces a new sandbox mode that allows `FaktoryWorker` to operate locally, without a Faktory server. This is intended to be used during testing, along with a handful of new test utilities contained in `FaktoryWorker.Testing`. The new `sandbox-testing` guide does a good job of summarizing:

---

# Sandbox Testing

When testing your application, you may not want to actually push/pull jobs to
Faktory, but instead rely on asserting that jobs _would be_ enqueued as you
expect. Enabling sandbox mode will accomodate this; when sandbox mode is active,
the following changes are made:

- no connection will be made to Faktory (you don't even need a Faktory server running!)
- jobs enqueued (with `perform_async/2`) will **never** be run
- jobs enqueued will be recorded, and can be viewed using `FaktoryWorker.Sandbox`

To enable sandbox mode, set `:sandbox` to `true`:

```elixir
# config/test.exs

config :my_app, FaktoryWorker, sandbox: true
```

```elixir
# lib/my_app/application.ex

defmodule MyApp.Application do
  @moduledoc false

  use Application

  def start(_type, _args) do
    children = [
      {FaktoryWorker, Application.get_env(:my_app, FaktoryWorker)}
    ]

    opts = [name: MyApp.Supervisor, strategy: :one_for_one]
    Supervisor.start_link(children, opts)
  end
end
```

You can verify that sandbox mode is enabled with `FaktoryWorker.Sandbox.active?/0`.

## Writing Tests

To aid in tests that rely on jobs being enqueued, as well as tests for job modules
themselves, you can use the helpers in `FaktoryWorker.Testing`. Below is a simple
test for a module that enqueues a job:

```elixir
defmodule MyApp.QueueTest do
  use ExUnit.Case

  import FaktoryWorker.Testing

  setup do
    reset_queues()
  end

  test "perform/1" do
    MyApp.Job.perform_async("hello, world!")
    assert_enqueued MyApp.Job, args: ["hello, world!"]
    refute_enqueued MyApp.Job, args: ["goodbye!"]
  end

  test "perform/2" do
    MyApp.Job.perform_async(["foo", "bar"])
    assert_enqueued MyApp.Job, args: ["foo", "bar"]
  end

  test "with opts" do
    MyApp.Job.perform_async("howdy", reserve_for: 1_500)
    assert_enqueued MyApp.Job, opts: [reserve_for: 1_500]

    # these would work too!
    # assert_enqueued MyApp.Job, args: ["howdy"]
    # assert_enqueued MyApp.Job, args: ["howdy"], opts: [reserve_for: 1_500]
  end
end
```

Here's a simple example test for a job module itself:

```elixir
defmodule MyApp.JobTest do
  use ExUnit.Case

  import FaktoryWorker.Testing

  test "handles strings" do
    assert perform_job(MyApp.Job, "foo") == :ok
  end

  test "handles maps" do
    assert perform_job(MyApp.Job, %{id: 1}) == {:ok, "cool, a map!"}
  end

  test "handles structs" do
    # structs are serialized and passed to `perform` as bare maps
    assert perform_job(MyApp.Job, %MyApp.User{id: 1}) == {:ok, "cool, a map!"}
  end

  test "doesn't handle numbers" do
    assert perform_job(MyApp.Job, 1_234) == {:error, "boo, no numbers!"}
  end

  test "gracefully handles raises" do
    assert perform_job(MyApp.Job, "I'll cause a raise!") == {:error, "raise reason"}
  end
end
```

When you can, prefer to use `perform_job/3` instead of calling `perform` directly, as
it will ensure that the code under test is run in the same way that it will eventually
be at runtime.

---

Closes #124.

superhawk610 commented:
I've created a minimal demo repo for testing out these changes [here](https://github.com/superhawk610/faktory_demo/tree/main/test).

swelham commented:
Thanks so much for the detailed write up both here and in the linked issue 🙌 

I have wanted to properly support testing for some time now and this approach looks great. I will take a look through the code and try it out with the test repo you provided, thanks for that also 👍 .

swelham commented on lib/faktory_worker/testing.ex:
```diff
@@ -0,0 +1,288 @@
+    #{describe_available_jobs(worker_mod)}
+    """
+
+    refute worker_mod |> Sandbox.find_jobs(filters) |> Enum.empty?(), err_message
```
By default I think this would ideally be asserting only a single job exists that matches the `worker_mod` and `filters`. This would catch bugs where duplicate jobs are getting enqueued in the case where it isn't expected.

To support the case where duplicate jobs are expected we could maybe support passing options into the assertion that indicate how many jobs are expected. Maybe something like

```elixir
assert_enqueued(MyJob, args: [123], count: 2)
```

Any thoughts?

superhawk610 replied:
I really like this idea. I'll also support passing `count: :any` to revert to the "at least one" behaviour, but leave a note in the docs that explicit counts should be preferred when possible.

swelham reviewed (commented):
This is looking great so far 👍 

After trying it out I have some initial suggestions on a few potential tweaks, but overall I really like the approach and I think this would be a great improvement to the library.

superhawk610 commented:
Thanks for the feedback, I really like all of your suggestions. I've updated the PR to incorporate them, and updated the documentation accordingly. Let me know what you think!
